### ISSUE TITLE
TaskLink hashCode and equals ignores 'type' and 'index' components

### ISSUE THREAD

**Issue by automenta** [opennars/opennars#550]
https://github.com/opennars/opennars/blob/0f446f737ebf36e3df58ac10050a71fb04f2b9ee/src/main/java/org/opennars/entity/TaskLink.java#L119

merges unpredictably on Bag nameTable put due to HashMap semantics.  

if this is somehow desired functionality, then an explicit random selection strategy for 'type' and 'index' and removal of the fields would be more clear - however I doubt that is the case.

a fix should include a unit test for these conditions.

similar defective hashcode/equality patterns may be present in other project entity types.

**Comment by automenta** [comment]
```java
public TaskLink(...
   ...
        this.hash = (((targetTask.hashCode() * 31) + type) * 31) + (index!=null ? Arrays.hashCode(index) : 0);
    }

    @Override
    public final int hashCode() {
        return hash;
    }
    
    @Override
    public final boolean equals(final Object obj) {
        if (obj == this) return true;
        if (obj instanceof TaskLink) {
            final TaskLink t = (TaskLink)obj;
            return hash==t.hash && type==t.type && Objects.equals(index, t.index) && targetTask.equals(t.targetTask);
        }
        return false;
    }
```

**Comment by patham9** [comment]
For a specific concept it is clear whether another term is a superterm etc, so there shouldn't be duplicates caused by type. For index it can be an issue as a subterm can appear twice. Let's improve this, thanks for finding!

@PtrMan: Do you think we can get Q/A metrics functioning for future improvements? Relevant would be answer time (penalizing late answers), answer truth value (penalizing not using available evidence) etc.

**Comment by automenta** [comment]
whatever fields are redundant (ie. can be inferred easily from context) should be removed for this reason. if a field serves only as cached value it should be clear so it doesnt get mixed in as a hashcode component and equality condition

**Comment by patham9** [comment]
Fixed in v3.0.3_otherFixes
This branch will finalize v3.0.3 for which some exceptions are left to fix.
